gtk4/src/subclass/widget.rs

```diff
@@ -983,6 +983,75 @@ pub unsafe trait WidgetClassSubclassExt: ClassStruct {
         }
     }
 
+    #[doc(alias = "gtk_widget_class_add_binding")]
+    fn add_binding<
+        F: Fn(&<<Self as ClassStruct>::Type as ObjectSubclass>::Type, Option<&Variant>) -> bool
+            + 'static,
+    >(
+        &mut self,
+        keyval: gdk::keys::Key,
+        mods: gdk::ModifierType,
+        callback: F,
+        arguments: Option<&glib::Variant>,
+    ) {
+        let shortcut = crate::Shortcut::new(
+            Some(&crate::KeyvalTrigger::new(keyval, mods)),
+            Some(&crate::CallbackAction::new(move |widget, args| -> bool {
+                unsafe { callback(widget.unsafe_cast_ref(), args) }
+            })),
+        );
+        shortcut.set_arguments(arguments);
+        self.add_shortcut(&shortcut);
+    }
+
+    #[doc(alias = "gtk_widget_class_add_binding_action")]
+    fn add_binding_action(
+        &mut self,
+        keyval: gdk::keys::Key,
+        mods: gdk::ModifierType,
+        action_name: &str,
+        arguments: Option<&glib::Variant>,
```

Comment:
Shouldn't this get some checks to see whether the action exists and has the correct parameter type(s)?

Reply:
After asking in #gtk channel, we can't really check whether the action exists as it can be registered later on. It can even be defined in a different action group that's added to this widget. Which makes the validation much more complex than what I imagined. I added a TODO comment about that for now